File: BinaryTrees/leetcode-543.py
#for a given tree: diameter = (1 + left sub tree height) + (1 + right sub tree height)
#find the diameter for every tree and update max_diameter
# How do we find the height? 
    #height = max(1 + left sub tree height, 1 + right sub tree height )
    #height = 0 for single node tree
    #height = -1 for empty tree

import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.max_diameter = 0

        def _height(root):
            if not root:
                return -1
            
            left_tree = _height(root.left)
            right_tree = _height(root.right)
            diameter =  1 + left_tree + 1 + right_tree
            if self.max_diameter < diameter:
                self.max_diameter = diameter
            return 1 + max(left_tree, right_tree)
        _height(root)
        return self.max_diameter




        

File: BinaryTrees/leetcode-543.cpp
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
   TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
   TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};


#include <bits/algorithmfwd.h>

class Solution {
public:
    int diameter = 0;
    
    // Height is defined as the number of edges from the node to its deepest leaf
    int height(TreeNode* node) {
        if (node == nullptr) {
            return 0; // Base case: no edges in an empty tree
        }
        
        // Recursively find the height of left and right subtrees
        int left_height = height(node->left);
        int right_height = height(node->right);
        
        // Update the diameter if the path through this node is longer
        diameter = std::max(diameter, left_height + right_height);
        
        // Height of the current node is 1 more than the maximum of its children's heights
        return 1 + std::max(left_height, right_height);
    }
    
    int diameterOfBinaryTree(TreeNode* root) {
        height(root);
        return diameter; // Diameter is the maximum number of edges found
    }
};
